colibre/scripts/bh_luminosities.py

```diff
@@ -0,0 +1,172 @@
+    bin_width = (np.log10(mass_bounds[1]) - np.log10(mass_bounds[0])) / bins
+
+    threshold_mass = 10 ** (
+        np.log10(mass_bins[mass_bins < masses_3rd_most_massive][-1]) + bin_width * 0.5
+    )
+    mass_bins = mass_bins[mass_bins < threshold_mass]
+
+    values_most_massive = values[masses > threshold_mass]
+    masses_most_massive = masses[masses > threshold_mass]
+
+    values_rest = values[masses < threshold_mass]
+    masses_rest = masses[masses < threshold_mass]
+
+    medians, _, _ = stats.binned_statistic(
+        masses_rest, values_rest, statistic="median", bins=mass_bins
+    )
+    percentile_10s, _, _ = stats.binned_statistic(
+        masses_rest,
+        values_rest,
+        statistic=lambda x: np.percentile(x, 10.0),
+        bins=mass_bins,
+    )
+    percentile_90s, _, _ = stats.binned_statistic(
+        masses_rest,
+        values_rest,
+        statistic=lambda x: np.percentile(x, 90.0),
+        bins=mass_bins,
+    )
+    mass_bins = np.array(
+        [(mass_bins[i] + mass_bins[i + 1]) / 2 for i in range(np.size(medians))]
+    )
+
+    return (
+        masses_most_massive,
+        values_most_massive,
+        masses_rest,
+        values_rest,
+        mass_bins,
+        medians,
+        percentile_10s,
+        percentile_90s,
+    )
+
+
+def make_single_image(
+    filenames, names, mass_bounds, value_bounds, number_of_simulations, output_path
+):
+
+    fig, ax = plt.subplots()
+
+    ax.set_xlabel("Black Hole Subgrid Masses $M_{\\rm sub}$ [M$_\odot$]")
+    ax.set_ylabel("Black Hole Luminosities $L_{\\rm bol}$ $[{\\rm erg}{\\rm s}^{-1}]$")
+    ax.set_xscale("log")
+    ax.set_yscale("log")
+
+    for filename, name in zip(filenames, names):
+        masses_most_massive, values_most_massive, masses_rest, values_rest, mass_bins, medians, percentile_10s, percentile_90s = calculate_medians(
+            filename, mass_bounds, value_bounds, bins
+        )
+        fill_plot, = ax.plot(mass_bins, medians, label=name)
+        ax.fill_between(
+            mass_bins,
+            percentile_10s,
+            percentile_90s,
+            alpha=0.2,
+            facecolor=fill_plot.get_color(),
+        )
+        scatter_plot = ax.scatter(
+            masses_most_massive, values_most_massive, facecolor=fill_plot.get_color()
+        )
+        if number_of_simulations == 1 and plot_scatter == True:
+
+            kwargs = dict(
+                edgecolor="none", zorder=-100, facecolor=fill_plot.get_color()
+            )
+
+            # Need to "intelligently" size the markers
+            kwargs["s"] = (
+                rcParams["lines.markersize"]
+                * (6.0 - 5.0 * np.tanh(0.75 * np.log10(masses_rest.size) - 3.0))
+                / 11.0
+            )
+
+            kwargs["alpha"] = (
+                5.5 - 4.5 * np.tanh(0.75 * np.log10(masses_rest.size) - 3.0)
+            ) / 10.0
+
+            ax.scatter(masses_rest, values_rest, **kwargs)
```

Comment:
Thanks for letting me know about these functions in velociraptor-python, I didn't know about them.

I have now implemented that function in all of the scripts (including the jet-related ones). The function does get called - I have checked. The plot is also created, but none of the dots actually appear on the plot, for some reason.

Reply:
Thanks @FHusko ! 

The dots were not showing up because `scatter_x_against_y()` would create its own `(fig,ax)` object.

I created an easy fix to this: https://github.com/SWIFTSIM/velociraptor-python/pull/93. If @bwvdnbro approves it, then in your scripts you can simply write 

```
if number_of_simulations == 1 and plot_scatter:
    scatter_x_against_y(ax, masses, values)
ax.scatter(additional_x.value, additional_y.value, color=fill_plot.get_color())
```

The `scatter_x_against_y` will take care of the dots, while the last line will highlight the N most massive objects. I have just checked that it all works

![black_hole_luminosities](https://user-images.githubusercontent.com/20153933/193430622-7efd0809-3e41-413f-b6ce-2d6044465545.png)

Reply:
https://github.com/SWIFTSIM/velociraptor-python/pull/93 has now been merged.

Reply:
Great, thanks both of you. I have updated all of the scripts with this (albeit I wasn't able to test it, with cosma being down).

I have also updated the flamingo scripts with all of the changes @EvgeniiChaikin suggested for the colibre ones so far. The only difference is that I turn off the scatter plots there by default.